Build fails with clang >= 22

Build errors:

```
/builddir/build/BUILD/fex-emu-2509.1-build/FEX-FEX-2509.1/ThunkLibs/Generator/analysis.cpp:316:103: error: attempt to use a deleted function
  316 |           auto repack_info_it = types.emplace(context.getCanonicalType(annotated_member->getParent()->getTypeForDecl()), RepackedType {}).first;
      |                                                                                                       ^
/usr/include/clang/AST/Decl.h:3982:15: note: 'getTypeForDecl' has been explicitly marked deleted here
 3982 |   const Type *getTypeForDecl() const = delete;
      |               ^
/builddir/build/BUILD/fex-emu-2509.1-build/FEX-FEX-2509.1/ThunkLibs/Generator/analysis.cpp:501:50: error: no member named 'getTypePtr' in 'clang::Type'
  501 |             types.emplace(context.getSizeType()->getTypePtr(), RepackedType {});
      |                           ~~~~~~~~~~~~~~~~~~~~~~~^
2 errors generated.
```

It looks like the first error was because of this change in clang: https://github.com/llvm/llvm-project/commit/91cdd35008e9ab32dffb7e401cdd7313b3461892

--- a/ThunkLibs/Generator/analysis.cpp
+++ b/ThunkLibs/Generator/analysis.cpp
@@ -313,7 +313,12 @@ void AnalysisAction::ParseInterface(clang::ASTContext& context) {
           }
 
           // Get or add parent type to list of structure types
+#if CLANG_VERSION_MAJOR >= 22
+          auto parent_qt = context.getTagType(clang::ElaboratedTypeKeyword::None, std::nullopt, annotated_member->getParent(), false);
+          auto repack_info_it = types.emplace(context.getCanonicalType(parent_qt).getTypePtr(), RepackedType {}).first;
+#else
           auto repack_info_it = types.emplace(context.getCanonicalType(annotated_member->getParent()->getTypeForDecl()), RepackedType {}).first;
+#endif
           if (repack_info_it->second.assumed_compatible) {
             throw report_error(template_arg_loc, "May not annotate members of opaque types");
           }
@@ -498,7 +503,7 @@ void AnalysisAction::ParseInterface(clang::ASTContext& context) {
             // Convert variadic argument list into a count + pointer pair
             data.param_types.push_back(context.getSizeType());
             data.param_types.push_back(context.getPointerType(*annotations.uniform_va_type));
-            types.emplace(context.getSizeType()->getTypePtr(), RepackedType {});
+            types.emplace(context.getSizeType().getTypePtr(), RepackedType {});
             if (!annotations.uniform_va_type.value()->isVoidPointerType()) {
               types.emplace(annotations.uniform_va_type->getTypePtr(), RepackedType {});
             }

--- a/ThunkLibs/Generator/gen.cpp
+++ b/ThunkLibs/Generator/gen.cpp
@@ -3,6 +3,7 @@
 #include "diagnostics.h"
 #include "interface.h"
 #include <clang/Frontend/CompilerInstance.h>
+#include <clang/Basic/DiagnosticOptions.h>
 
 #include <fstream>
 #include <numeric>
@@ -816,7 +817,10 @@ bool GenerateThunkLibsActionFactory::runInvocation(std::shared_ptr<clang::Compil
 
   GenerateThunkLibsAction Action(libname, output_filenames, abi);
 
-#if LLVM_VERSION_MAJOR >= 20
+#if LLVM_VERSION_MAJOR >= 22
+  auto Diags = clang::CompilerInstance::createDiagnostics(Compiler.getVirtualFileSystem(), Compiler.getDiagnosticOpts(), DiagConsumer, false);
+  Compiler.setDiagnostics(std::move(Diags));
+#elif LLVM_VERSION_MAJOR >= 20
   Compiler.createDiagnostics(Compiler.getVirtualFileSystem(), DiagConsumer, false);
 #else
   Compiler.createDiagnostics(DiagConsumer, false);
@@ -825,7 +829,11 @@ bool GenerateThunkLibsActionFactory::runInvocation(std::shared_ptr<clang::Compil
     return false;
   }
 
+#if LLVM_VERSION_MAJOR >= 22
+  Compiler.createSourceManager();
+#else
   Compiler.createSourceManager(*Files);
+#endif
 
   const bool Success = Compiler.ExecuteAction(Action);
 

--- a/unittests/ThunkLibs/abi.cpp
+++ b/unittests/ThunkLibs/abi.cpp
@@ -116,7 +116,10 @@ public:
     Compiler.setInvocation(std::move(invocation));
 #endif
     Compiler.setFileManager(files);
-#if LLVM_VERSION_MAJOR >= 20
+#if LLVM_VERSION_MAJOR >= 22
+    auto Diags = clang::CompilerInstance::createDiagnostics(Compiler.getVirtualFileSystem(), Compiler.getDiagnosticOpts(), diag_consumer, false);
+    Compiler.setDiagnostics(std::move(Diags));
+#elif LLVM_VERSION_MAJOR >= 20
     Compiler.createDiagnostics(Compiler.getVirtualFileSystem(), diag_consumer, false);
 #else
     Compiler.createDiagnostics(diag_consumer, false);
@@ -124,7 +127,12 @@ public:
     if (!Compiler.hasDiagnostics()) {
       return false;
     }
+
+#if LLVM_VERSION_MAJOR >= 22
+    Compiler.createSourceManager();
+#else
     Compiler.createSourceManager(*files);
+#endif
 
     const bool Success = Compiler.ExecuteAction(ScopedToolAction);
 

